source/frontend/editors/RarcEditor.cpp

```diff
@@ -0,0 +1,1035 @@
+                            std::vector<std::filesystem::path>& out) {
+  std::vector<std::filesystem::path> dirs;
+  for (auto&& it : std::filesystem::directory_iterator{path}) {
+    if (it.is_directory())
+      dirs.push_back(it.path());
+    else
+      out.push_back(it.path());
+  }
+  for (auto& dir : dirs) {
+    out.push_back(dir);
+    get_sorted_directory_list_r(dir, out);
+  }
+}
+
+static void renameNode(ResourceArchive& rarc,
+                       std::optional<ResourceArchive::Node>& node,
+                       const std::string& new_name) {
+  RSL_DEFER(node = std::nullopt);
+
+  if (!node)
+    return;
+
+  if (new_name == "")
+    return;
+
+  auto node_it = std::find(rarc.nodes.begin(), rarc.nodes.end(), *node);
+  if (node_it == rarc.nodes.end())
+    return;
+
+  node_it->name = new_name;
+}
+
+static bool insertFiles(ResourceArchive& rarc,
+                        std::optional<ResourceArchive::Node> parent,
+                        std::vector<rsl::File>& files) {
+  RSL_DEFER(files.clear());
+
+  if (files.size() == 0)
+    return false;
+
+  if (!parent)
+    return false;
+
+  std::vector<ResourceArchive::Node> new_nodes;
+  for (auto& file : files) {
+    auto file_name = file.path.filename().string();
+    std::transform(file_name.begin(), file_name.end(), file_name.begin(),
+                   [](u8 ch) { return std::tolower(ch); });
+    ResourceArchive::Node node = {.id = 0,
+                                  .flags = ResourceAttribute::FILE |
+                                           ResourceAttribute::PRELOAD_TO_MRAM,
+                                  .name = file_name,
+                                  .data = file.data};
+    new_nodes.push_back(node);
+  }
+
+  int parent_index =
+      std::distance(rarc.nodes.begin(),
+                    std::find(rarc.nodes.begin(), rarc.nodes.end(), *parent));
+
+  // Manual sort is probably better because we can dodge DFS nonsense
+  // effectively Even with O(n^2) because there typically aren't many nodes
+  // anyway.
+  for (auto new_node = new_nodes.begin(); new_node != new_nodes.end();
+       new_node++) {
+    auto dir_files_start = rarc.nodes.begin() + parent_index + 1;
+    auto dir_files_end = rarc.nodes.begin() + parent->folder.sibling_next +
+                         std::distance(new_nodes.begin(), new_node);
+    for (auto child = dir_files_start; /*nothing*/; child++) {
+      // Always insert at the very end (alphabetically last)
+      if (child == dir_files_end) {
+        rarc.nodes.insert(child, *new_node);
+        break;
+      }
+      if (child->is_folder()) {
+        // Never insert before the special dirs
+        if (child->is_special_path())
+          continue;
+        // Always insert before the regular dirs
+        rarc.nodes.insert(child, *new_node);
+        break;
+      }
+      if (new_node->name < child->name) {
+        rarc.nodes.insert(child, *new_node);
+        break;
+      }
+    }
+  }
```

Comment:
Would this make more sense in library code?

Reply:
Probably. At the time of writing I had a mentality of the operations being less to do with the RARC itself and more just surgical operations for the editor but that would make more sense.